### ISSUE TITLE
Publish a textmate gem

### ISSUE THREAD

**Issue by jeff-hykin** [jeff-hykin/cpp-textmate-grammar#290]
I'm creating this as a checklist for publishing the gem
- create a Pattern class instead of having newPattern use regex
- use the turnOffNumberedCaptureGroups on every regex pattern
- add a check for newlines, issue a warning
- switch to snake_case for everything
- check/fix the `tag_as:` in oneOrMoreOf() or zeroOrMoreOf()
- add a `to_s` method to Pattern and PatternRange
- retagging feature
- create an "export_all" which exports the language tags, json syntax, and yaml syntax to a folder
- create a standard library of patterns
    - numerics
    - block and parentheses finder
    - strings and escapes
    - inline comments
    - standard space
- add documentation to a gitbook
  - The Grammar class
    - init arguments
    - $initial_context
    - setting repos
    - exporting
    - the named patterns (@spaces, @word)
  - The Token class
    - That
      - using `not()`
    - RepresentationsThat
  - The Pattern class
    - init arguments
      - backreferences
      - $match
      - $reference
    - all of the helper methods (then, or, zeroOrMoreOf, oneOrMoreOf, etc)
  - Pattern Range
    - init arguments
  - The standard library

**Comment by matter123** [comment]
We may want to reevaluate generating `*.tmLanguage.yaml` files. With, afaik, no text editor or tool supporting yaml language, and with the JSON files being pretty-printed, I not sure what their purpose is.

**Comment by jeff-hykin** [comment]
yeah. I think its safe to remove them, I only added yaml because I didn't pretty print the json.

**Comment by matter123** [comment]
Fairly minor concern but 
> create a Pattern class instead of having newPattern use regex

would prevent `/foo/.maybe(bar)` from working, correct?

If so, there are 33 instances of that that would need to be replaced.

**Comment by jeff-hykin** [comment]
Yeah, and that was the original reason that the pattern class was never created. If I made it a class the idea would be to clean everything up and let it be symbolic, and then only convert it to regex once `to_tag` was called. But I'm not sure it's worth all the trouble.

**Comment by jeff-hykin** [comment]
I've also been debating the snake case switch. I prefer static data to be sake case and functions to be camel, but Ruby uses snake for both. If I switched it would only be for formatting consistency.

**Comment by matter123** [comment]
It seems the majority of the uses for starting with a regexp is looking for a literal (17), with anchors being the nxt largest (6).

**Comment by matter123** [comment]
The warning for newlines is probably too overly broad to be useful. In particular, the new include pattern trips up my initial warning.

Output:
```
There is a pattern that likely tries to match characters after \n
textmate grammars only operate on a single line, \n is the last possible character that can be matched.
Here is the pattern:
((?:^)((?:(?:(?>\s+)|(\/\*)((?>(?:[^\*]|(?>\*+)[^\/])*)((?>\*+)\/)))+?|(?:(?:(?:(?:\b|(?<=\W))|(?=\W))|\A)|\Z)))((#)\s*((?:(?:include|include_next)|import))\b)\s*(?:(?:(?:((<)[^>\n]*(>?)((?:(?:(?>\s+)|(\/\*)((?>(?:[^\*]|(?>\*+)[^\/])*)((?>\*+)\/)))+?|(?:(?:(?:(?:\b|(?<=\W))|(?=\W))|\A)|\Z)))(?:(?:\n|$)|(?=\/\/)))|((\")[^\"\n]*(\"?)((?:(?:(?>\s+)|(\/\*)((?>(?:[^\*]|(?>\*+)[^\/])*)((?>\*+)\/)))+?|(?:(?:(?:(?:\b|(?<=\W))|(?=\W))|\A)|\Z)))(?:(?:\n|$)|(?=\/\/))))|((?:[a-zA-Z_]|(?:\\u[0-9a-fA-F]{4}|\\U[0-9a-fA-F]{8}))(?:[a-zA-Z0-9_]|(?:\\u[0-9a-fA-F]{4}|\\U[0-9a-fA-F]{8}))*((?:(?:(?>\s+)|(\/\*)((?>(?:[^\*]|(?>\*+)[^\/])*)((?>\*+)\/)))+?|(?:(?:(?:(?:\b|(?<=\W))|(?=\W))|\A)|\Z)))(?:(?:\n|$)|(?=\/\/))))|((?:(?:(?>\s+)|(\/\*)((?>(?:[^\*]|(?>\*+)[^\/])*)((?>\*+)\/)))+?|(?:(?:(?:(?:\b|(?<=\W))|(?=\W))|\A)|\Z)))(?:(?:\n|$)|(?=\/\/))))
```

Newline detection:
```ruby
# check for matching after \n
if /\\n(.*?)(?=!\||\\n)/ =~ regex_as_string
    if /[^\^$\[\]\(\)?:+*=!<>\\]/ =~ $1
        
        puts "\n\nThere is a pattern that likely tries to match characters after \\n\n"
        puts "textmate grammars only operate on a single line, \\n is the last possible character that can be matched.\n"
        puts "Here is the pattern:\n"
        puts regex_as_string
        # raise "Error: see printout above"
    end
end
```

**Comment by matter123** [comment]
Does
> check/fix the tag_as: in oneOrMoreOf() or zeroOrMoreOf()

Refer to the fact that the pattern below will only give the last `foo` the scope of `test.foo`? If so, I think adding a warning about that would harm usability, composing patterns should be able to work without having to call turn_of_numbered_capture_groups whenever you want to match the pattern multiple times.

```ruby
grammar[:test] = newPattern(
    match: oneOrMoreOf(newPattern(match:/foo/, tag_as:"test.foo").maybe(@spaces)),
    tag_as: "test"
)
```

**Comment by jeff-hykin** [comment]
> Does
> 
> > check/fix the tag_as: in oneOrMoreOf() or zeroOrMoreOf()
> 
> Refer to the fact that the pattern below will only give the last `foo` the scope of `test.foo`? If so, I think adding a warning about that would harm usability, composing patterns should be able to work without having to call turn_of_numbered_capture_groups whenever you want to match the pattern multiple times.
> 
> ```ruby
> grammar[:test] = newPattern(
>     match: oneOrMoreOf(newPattern(match:/foo/, tag_as:"test.foo").maybe(@spaces)),
>     tag_as: "test"
> )
> ```

Yeah that was the problem I was referring to. I was thinking of automatically disabling the tags that are inside the oneOrMoreOf() or zeroOrMoreOf() and then automatically adding an `includes:` that has the repeated pattern one time, which I believe will then correctly tag all of the repetitions.

**Comment by jeff-hykin** [comment]
> ```ruby
> \\n(.*?)(?=!\||\\n)
> ```

I was thinking it would be a warning just like the zero-length pattern. Its given, but it can be disabled because sometimes it's valid.

That said I'm not sure why there's a `[^>\n]` inside the pattern. There's other \n's too but `[^>\n]` should just be `[^>]`

**Comment by jeff-hykin** [comment]
> Fairly minor concern but
> 
> > create a Pattern class instead of having newPattern use regex
> 
> would prevent `/foo/.maybe(bar)` from working, correct?
> 
> If so, there are 33 instances of that that would need to be replaced.

We could just do
```
class Pattern < Regexp
    def initialize(arguments)
        self.then(arguments)
    end
end
```
That we could use `Pattern.new` in the guides/explanations to be consistent with PatternRange.new()

**Comment by jeff-hykin** [comment]
I think theres no way to do the `use the turnOffNumberedCaptureGroups on every regex pattern` without having a refactored Pattern class. Right now there's no way to fully tell the difference between pure regex and pattern-combined regex.

The next closest thing would be to count the number of capture groups there should be, and the number of capture groups their actually are.

**Comment by matter123** [comment]
> That we could use Pattern.new in the guides/explanations to be consistent with PatternRange.new()

That sounds good.

> I was thinking it would be a warning just like the zero-length match warning. Its given, but it can be disabled because sometimes it's valid.

> That said I'm not sure why there's a [^>\n] inside the pattern. There's other \n's too but [^>\n] should just be [^>]

I'll clean up the include pattern, add the ability to disable the warning and push that branch.

**Comment by matter123** [comment]
Another check that might be worth considering is wrong ordering of patterns with a common prefix.

If a pattern set contains the following patterns:
```regex
...
[A-Z_]+       # pattern 1
[A-Z_]+\s+foo # pattern 2
...
```
Pattern 2 will never match because it has a common prefix with pattern 1 and pattern 1 is first.

**Comment by jeff-hykin** [comment]
> Another check that might be worth considering is wrong ordering of patterns with a common prefix.
> 
> If a pattern set contains the following patterns:
> 
> ```regexp
> ...
> [A-Z_]+       # pattern 1
> [A-Z_]+\s+foo # pattern 2
> ...
> ```
> 
> Pattern 2 will never match because it has a common prefix with pattern 1 and pattern 1 is first.

Do you think this is detectable? I guess check for a perfect subset at the beginning of two patterns.

**Comment by matter123** [comment]
Yeah, for any pattern A of the pattern set 1..N:
	for each pattern B in the range 1..A-1:
		if B is a prefix of A, issue a warning/error

**Comment by matter123** [comment]
the `zeroLengthStart` is a bit over-cautious, in particular, \G is usually intentional, and as such \G should not be enough to trigger the warning.

One possible solution is to do what vscode-textmate does when attempting to not evaluate \G, that is, replace \G with \uFFFF.

**Comment by matter123** [comment]
Depending on whether you plan on supporting generating atom themes, you may want to split tag_as by the space character and generate multiple captures. See https://github.com/microsoft/vscode-textmate/issues/108#issuecomment-522197256

**Comment by matter123** [comment]
It's somewhat of an odd case, but should the grammar generator detect when you are using a start of line anchor inside a sub match, and then rewrite it so that it works?.

Example:
Orignal (non working):
```
Pattern.new(
  match: /\W+/,
  includes: [
    Pattern.new(match: /^bar/, tag_as: "bar")
  ]
)
```
Rewritten (working)
```
Pattern.new(
  match: /\W+/,
  includes: [
    PatternRange.new(
      start_pattern: lookAheadFor(/./),
      end_pattern: /$/,
      includes: [
        Pattern.new(match: /^bar/, tag_as: "bar")
      ]
    )
  ]
)
```

**Comment by jeff-hykin** [comment]
> Depending on whether you plan on supporting generating atom themes, you may want to split tag_as by the space character and generate multiple captures. See [microsoft/vscode-textmate#108 (comment)](https://github.com/microsoft/vscode-textmate/issues/108#issuecomment-522197256)

I think this is a perfect situation for the library to handle (to support Atom), but it is low priority so we should put the effort in after the gem is published.

**Comment by jeff-hykin** [comment]
> It's somewhat of an odd case, but should the grammar generator detect when you are using a start of line anchor inside a sub match, and then rewrite it so that it works?.
> 
> Example:
> Orignal (non working):
> 
> ```
> Pattern.new(
>   match: /\W+/,
>   includes: [
>     Pattern.new(match: /^bar/, tag_as: "bar")
>   ]
> )
> ```
> 
> Rewritten (working)
> 
> ```
> Pattern.new(
>   match: /\W+/,
>   includes: [
>     PatternRange.new(
>       start_pattern: lookAheadFor(/./),
>       end_pattern: /$/,
>       includes: [
>         Pattern.new(match: /\Gbar/, tag_as: "bar")
>       ]
>     )
>   ]
> )
> ```
> 
> It is fairly unintuitive that /^/ doesn't work inside sub captures. And equally unituitive that the solution is to wrap it up in a PatteernRange.

Yeah I think this is another perfect case for the library to handle. Like with the Atom support though, handle it after the gem is published

**Comment by matter123** [comment]
Since quite a few check items and potential additions perform transformations on the patterns, there should be some way of enabling logging whenever a transformation occurs.

**Comment by jeff-hykin** [comment]
I think I'm going to try to publish this as a pre-alpha gem sometime this week. Its the first week of the semester and I want to try to re-structure this repo (get each language in a separate repo, and coordinate the change with Alexr00) before the semester gets into full swing

The only thing I want to change before that is the import/export method. I think I'm going to change it to something like

```ruby
require 'textmate_tools'

grammar = Grammer.get_exportable_grammar
grammar.exports = [
    :repo_name_1,
    :repo_name_2,
]
grammar.external_repos = [
    :std_space,
    :variable,
]

grammar[:repo_1] = Pattern.new()
```

The `.exports` are the names of the repos that should be exported
The `.external_repos` are the ones you expect to already be defined by something else
Every time that `grammar` sees a repo name (even in the includes of a pattern), if its not in the `.exports` and not in the `.external_repos`, it will create a mapping from that repo name to a randomly generated number so that the name never conflicts with a name from another package.

**Comment by matter123** [comment]
That sounds pretty nice.
I have two concerns:
1. The mapping should probably be stateless;
2. Rerunning `npm run build` should not change the generated grammar(builds should be reproducible)

so I'm thinking:
- Grammar#get_exportable_grammar returns a new ExportGrammar that contains a reproducible salt (hash of the *relative* filename, some "export name", etc")
 - whenever the ExportGrammar encounters a name that is not in exports or external_repos, the name is prepended with the salt

**Comment by jeff-hykin** [comment]
Yeah salt based on file name/path would be ideal

**Comment by matter123** [comment]
With a unified export, to another grammar or to a file, how does an import look like?

**Comment by matter123** [comment]
`source/rewrite/example_grammar.rb` contains an example import/export:

```ruby
require_relative 'grammar'

ex = Grammar.new_exportable_grammar

ex.exports = [
    :abc,
]

ex.external_repos = [
    :p123,
]

ex[:abc] = Pattern.new(
    match: /abc/,
    includes: [
        :p123,
        :test_pat,
        :$initial_context,
    ]
)

ex[:test_pat] = /abcd/

g = Grammar.new(
    "test",
    "source.test",
)

ex.export # just to confirm that export is idempotent

g.import(ex.export)

g.debug
```

**Comment by jeff-hykin** [comment]
> In theory, the rewrite has
> 
> * `Grammar::new_exportable_grammar` returns a new ExportableGrammar which is suitable for importing
> * `Grammar::fromTmLanguage(path)` - loads a json file as an ImportableGrammar which has additional safeguards on accessing hashes
> * `ExportableGrammar#export` preps the ExportableGrammar for exporting (stub)
> * `Grammar#import(path_or_export)`:
>   
>   * if `path_or_export` is an ExportableGrammar import the repository
>   * if `path_or_export` is a path, eval the file and import the ExportableGrammar

Yup, this sounds great 👍

**Comment by jeff-hykin** [comment]
> `source/rewrite/example_grammar.rb` contains an example import/export:
> 
> ```ruby
> require_relative 'grammar'
> 
> ex = Grammar.new_exportable_grammar
> 
> ex.exports = [
>     :abc,
> ]
> 
> ex.external_repos = [
>     :p123,
> ]
> 
> ex[:abc] = Pattern.new(
>     match: /abc/,
>     includes: [
>         :p123,
>         :test_pat,
>         :$initial_context,
>     ]
> )
> 
> ex[:test_pat] = /abcd/
> 
> g = Grammar.new(
>     "test",
>     "source.test",
> )
> 
> ex.export # just to confirm that export is idempotent
> 
> g.import(ex.export)
> 
> g.debug
> ```

Thanks for full hashing this out. I think this is pretty ideal. The only thing I might change is the `ex.export # just to confirm that export is idempotent`. A file might want to dynamically contribute values, like the `generateBlockFinder`. This should possible because they'll still have access to the exportable grammar object. We can give the exportable grammar access to the real grammar and let it inject repos dynamically.

With that being the case, I think we can have a class-variable `Grammar@@exportable_grammar_stack`. Whenever a exportable grammar is constructed, it places itself on that stack. When using `Grammar#import(path)` we can see how many exportable grammars are created by a particular file, and then warn them if there is no exportable grammar, or if there are multiple (ideally there would only be one, kind of similar the the Javascript `module.exports`)

**Comment by matter123** [comment]
Can you provide an example of what dynamically contributed values might look like?

**Comment by matter123** [comment]
Would `ExportGrammar#[]` where a symbol in external_repos is provided, return a PlaceholderPattern that during import is resolved to the actual pattern work?

For that matter should `Grammar#[]` just return a placeholder pattern whenever the key doesn't exist?

**Comment by jeff-hykin** [comment]
```
require_relative 'grammar'

ex = Grammar.new_exportable_grammar


ex.exports = [
    :functions
]
ex.external_repos = [
    :numbers,
]

# start as empty context
ex[:functions] = []

def generateFunctionFor(name)
    # range version
    ex[name.to_sym] = PatternRange.new(
        start_pattern: /#{name}\(/,
        end_pattern: /\)/,
        includes: [
            :numbers
        ]
    )
    # inline version
    ex[(name+"_inline").to_sym] = Pattern.new(/#{name}\(.+\)/)
    # add them to the function context that is exported
    ex[:functions] = [
        (name.to_sym),
        (name+"_inline").to_sym
    ]
end

g = Grammar.new(
    "test",
    "source.test",
)
g.import(ex)
```

**Comment by matter123** [comment]
To clarify the main generate file calls `generateFunctionFor()` right?

After ex is imported, g could assign itself as a parent Grammar, and at that point `[]` and `[]=` call the parent grammars functions.

Alternatively, `ex.reexport` could be added which basically does `g.import(ex)` again.

**Comment by jeff-hykin** [comment]
Yeah, all that sounds correct.

I think the way to do it will be for the exportable grammar's `[]=` to assign something like a PlaceholderPattern (maybe `ImportedPattern`) to the main grammar. It would be the same as a regular pattern, but it would contain a back-reference to the exportable grammar.

Repo names `ex[:repo_name]` that are not in `ex.exports` or `ex.external_repos`, should be converted on the spot to their hashed/no-conflict versions `ex[:repo_name]=` becomes `real_grammar[:some_hash]=`.

However, the Pattern.includes `Pattern.new(match: /thing/, includes:[ :repo_name ])` will need to be converted to `Pattern.new(match: /thing/, includes:[ :some_hash ])` once the real grammar is being exported to json. This is because things can get dynamically inserted into the `includes:`. We can probably have `ImportedPattern` inherit from Pattern, then override the `to_tag` to perform all of the replacements that need to be done on the includes.

**Comment by matter123** [comment]
Currently export uses `Pattern#transform_includes` to fixup its includes (see fixupValue on line 147 of grammar.rb)

**Comment by jeff-hykin** [comment]
👍 perfect, then the override can be done there. I think the full-grammar tests (making sure all repos exist) are the only things that would depend on the transforms.

**Comment by matter123** [comment]
To summarize then:
 - if the export grammar wants to inject dynamic patterns it should call ex.re_export
 - re_export does nothing if export has not been called, otherwise copies the new patterns to the parent grammar
- import sets self as the parent to the export

Thoughts on `Grammar#[]` returning a `PlaceholderPattern` rather than `nil`?

**Comment by jeff-hykin** [comment]
> if the export grammar wants to inject dynamic patterns it should call ex.re_export

Actually I was thinking by default the exportable grammar would be dynamic. Something like this:

```ruby
class Grammar
    @@exported_grammar_stack
    
    def import(filepath)
        before_size = @@exported_grammar_stack.size 
        require filepath
        after_size = @@exported_grammar_stack.size
        if before_size == after_size + 1
            exported_grammar = @@exported_grammar_stack.pop()
            # give the exported grammar access to the real grammar
            exported_grammar.attach_to_grammar(self)
        else
            raise "error: didn't export a grammar or exported more than 1 grammar in #{filepath}"
        end
        
    end
end


class ExportGrammar
    @real_grammar = []
    @repo = {}
    
    def initilize
        # tell all grammars that an exportable grammar was just made
        Grammar.exported_grammar_stack.push(self)
    end
    
    def [](key)
        key = self.hash_key_if_needed(key)
        return @repo[key]
    end
    
    def []=(key, value)
        pattern = ImportedPattern.new(value)
        key = self.hash_key_if_needed(key)
        @repo[key] = pattern
        if @real_grammar != nil
            @real_grammar[key] = pattern
        end
    end
    
    def attach_to_grammar(real_grammar)
        @real_grammar = real_grammar
        # copy over all current stuff
        for each_key, each_value in @repo
            @real_grammar[each_key] = each_value
        end
    end
end
```

**Comment by matter123** [comment]
Slight issue,
If the generate.rb requires a file that exports a grammar first when import is called there is no change to the @@exported_grammar_stack, `load` would work though, right?

**Comment by matter123** [comment]
The initialize for ExportGrammar could register with a key to its filename, that would allow for nested requires and still allow for an error with an unexpected number of grammars.

**Comment by jeff-hykin** [comment]
> Slight issue,
> If the generate.rb requires a file that exports a grammar first when import is called there is no change to the @@exported_grammar_stack, `load` would work though, right?

load would work. We could also change it from a stack to a Hash and then we could make sure the same file never gets imported twice

**Comment by matter123** [comment]
I am going to make it a Hash storing the seed, and use the line number of the call to determine if this is a different grammar in the same file, and if so, complain.

This allows, if needed, for `load` to be used, while still limiting it to one export/partial Grammar per file.

**Comment by jeff-hykin** [comment]
Okay, sounds good to me.

**Comment by jeff-hykin** [comment]
it'll probably be saturday before I make changes, but I think I'm going to pull out the other languages and have this repo only contain C++ and the TextMate tools. If we publish a gem, it is going to be a pain to test the library because it will need to be published, then pulled into C++ the tested. That or some more complicated workaround is going to be needed.

So I'll just leave the other languages frozen with an old version of the TextMate tools. That way this repo can make updates to the library and C++ until the library is pretty stable. Since I've already missed the deadline of school starting for me, I'm fine to focus more on C++, and hopefully get the library stable by Christmas. I think doxygen and a bailout for the macro conditionals are the next major changes

**Comment by matter123** [comment]
So this is a rough outline of how I imagined internal bailout would work.

Interface
```ruby
grammar.add_bailout(
  name: "macro",
  pattern: lookBehindToAvoid("\\").then("\n"),
  # or end_pattern, while_pattern if separate patterns are required
  priority: 1 # optional defaults to zero
)
```

when save_to is called, each bailout (sorted by priority then declaration):
  1. makes a copy of each PatternRange
  2. Adds those patterns and reference to any match patterns to `"#{bailout_name}_initial_context".to_sym`
  3. rewrite includes to refer to the bailed out patterns

**Comment by matter123** [comment]
Alternatively, any match patterns that are added to `:$initial_context` could instead be added to `match_initial_context` and then any initial_contexts could then just include that to save space.

**Comment by jeff-hykin** [comment]
I was thinking the bailout would return a hash (a textmate pattern) that utilizes the `"repository":` key to prevent name collisions. I haven't fully tested the  `"repository":` key though. Something like:
```ruby
grammar[:bailed_out_enum_context] = grammar.bailout(
    context: grammar[:enum_context], # an array of patterns
    pattern: lookBehindToAvoid("\\").then("\n"),
)
# returns something like
{
    repository: {
        # this is the bailed out enum_member, but it should be safe for it to have the same name 
        # because its in a repository 
        enum_member: {
            match: /stuff/
        }
    }
    includes: [
        # defaults (hopefully) to the closest defintion: aka the bailed out enum_member
        :enum_member,
    ]
}
```

**Comment by jeff-hykin** [comment]
Also I created repos for shell/dockerfile/perl. Not sure if seperate ones should be created for Objc, Objcpp, and C.
Anyways, those other languages can be safely deleted now and should help with compile time performance

**Comment by matter123** [comment]
That sounds reasonable, but we still want to defer generation  so that patterns added after are still bailed out. The fact that  `repository` is not documented anywhere is concerning, however.

**Comment by matter123** [comment]
The rewrite branch now has the rewrite in the top-level directory `gem`.

**Comment by jeff-hykin** [comment]
> all methods and classes in the directory gem/lib have at least some form of documentation

Awesome! yardoc sounds like the way to go.

Using "textmate-grammar" as the gem name works for me.

> so that patterns added after are still bailed out.

I was thinking baliout would be done inline (performed on-call) but that does present problems if it is called before some of the repo-names (e.g. `:comment` ) have been assigned to a pattern. ex:
```
macro_context = bailout(grammar[:$initial_context])
grammar[:comment] = Pattern.new()
```
This would make imported code (like preprocessor) less modular since they would need a secondary function to be called near the end of the grammar definition to setup the bailed-out code.

However, it could be difficult to do the bailouts in an automatic way at the end of a grammar export. If bailoutA includes[ :comment, :bailoutB ] and then bailoutB includes[ :preprocessor, :bailoutA]  then there would need to be 4 bailouts.
1. pure bailoutA (the top level one that's included within most things)
2. pure bailoutB ^
3. bailoutA + B 
4. bailoutB + A

the grammar uses the pure bailouts, but then all 4 of the bailouts would :include the A+B or B+A versions so that their internal patterns always closed early.

Since this case is pretty complicated, we could have two systems. 1. A manual bailout that gets called inline, leaving the complexity/order up to the user and then 2. A bailout that is a placeholder, which is converted to a real pattern when the grammar is exported to json. However, the bailout placeholder wouldn't be allowed to include (directly or indirectly) other bailout placeholders. 

> The fact that repository is not documented anywhere

Yeah... gotta love TextMate's documentation 🙃

**Comment by matter123** [comment]
My thought was the global bailouts would just be processed in order so that later bailout patterns (or of higher priority) would be able to bailout earlier ones but not Visa-Versa.
In effect 3 is folded into 2 and 4 doesn't exist.

Alternatively, there could just be a single global bailout pattern.

**Comment by matter123** [comment]
Thinking some more, I think bailouts can be done entirely outside of the Grammar object (assuming Grammar is enumerable).

`add_early_bailout_to_grammar(grammar, bailout_pattern, prefix)` would enumerate the repository and any pattern ranges would be duplicated, the bailout_pattern added, then stored as `grammar[(prefix+"_"+name.to_s).to_sym]`

So right before saving the generate file could call that.

More consistent with what we already have:
`Grammar.add_transform(Bailout.new(prefix, pattern))`

That performs basically the same steps as textmate_bailout

**Comment by jeff-hykin** [comment]
@matter123  Yesterday I spent the whole day trying to add `TokenPattern` as a kind of clone of `PlaceholderPattern`.  It inherited from `PlaceholderPattern` and just overloaded the `resolve` method, but running it caused a complex shared infinite recursion/loop. It was somehow related to __deep_clone__ so I started working on printing and simplifying the cloning logic, but that became pretty difficult as well. I think the bug was related to the deep cloning logic of Placeholder, insert, and __deep_clone_self__ but I never found it.

After I rewrote half of it, I understood the logic of the system and basically just made a change to the `resolve_placeholders.rb` transform. It expands out the `resolve` logic (resolve() is basically unused now), and adds the logic for resolving TokenHelpers.

Even though all tests pass, these changes shouldn't be on master so I'll put them on another branch and revert them later today. Once the code is cleaned up, I'll merge it back in.
